From f65e0ff12bfa0664938ba54332c26cef7d2c9d0d Mon Sep 17 00:00:00 2001
From: Herman van Hazendonk <github.com@herrie.org>
Date: Wed, 4 Jun 2025 16:22:59 +0200
Subject: [PATCH] wayland_touch: apply device_scale_factor to events

When the --force-device-scale-factor option is used, the touch coordinates in the
window need to be scaled accordingly before being passed on.

Upstream-Status: Pending

Signed-off-by: Christophe Chapuis <chris.chapuis@gmail.com>
---
 src/ui/ozone/platform/wayland/host/wayland_touch.cc | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/src/ui/ozone/platform/wayland/host/wayland_touch.cc b/src/ui/ozone/platform/wayland/host/wayland_touch.cc
index ef00f2a756..5740089732 100644
--- a/src/ui/ozone/platform/wayland/host/wayland_touch.cc
+++ b/src/ui/ozone/platform/wayland/host/wayland_touch.cc
@@ -11,6 +11,8 @@
 #include "base/time/time.h"
 #include "ui/events/types/event_type.h"
 #include "ui/gfx/geometry/point_f.h"
+#include "ui/display/display.h"
+#include "ui/display/screen.h"
 #include "ui/ozone/platform/wayland/common/wayland_util.h"
 #include "ui/ozone/platform/wayland/host/wayland_connection.h"
 #include "ui/ozone/platform/wayland/host/wayland_serial_tracker.h"
@@ -106,6 +108,10 @@ void WaylandTouch::OnTouchDown(void* data,
 #endif  // defined(OS_WEBOS)
   gfx::PointF location = self->connection_->MaybeConvertLocation(
       gfx::PointF(wl_fixed_to_double(x), wl_fixed_to_double(y)), window);
+
+  display::Display display = display::Screen::GetScreen()->GetPrimaryDisplay();
+  location.Scale(display.device_scale_factor());
+
   base::TimeTicks timestamp = base::TimeTicks() + base::Milliseconds(time);
   self->delegate_->OnTouchPressEvent(window, location, timestamp, id,
                                      EventDispatchPolicyForPlatform()
@@ -158,6 +164,10 @@ void WaylandTouch::OnTouchMotion(void* data,
   }
   gfx::PointF location = self->connection_->MaybeConvertLocation(
       gfx::PointF(wl_fixed_to_double(x), wl_fixed_to_double(y)), target);
+
+  display::Display display = display::Screen::GetScreen()->GetPrimaryDisplay();
+  location.Scale(display.device_scale_factor());
+
   base::TimeTicks timestamp = base::TimeTicks() + base::Milliseconds(time);
   self->delegate_->OnTouchMotionEvent(location, timestamp, id,
                                       EventDispatchPolicyForPlatform()
